src/utils/bindtwostepcarettoattribute.js

```diff
@@ -0,0 +1,143 @@
+/**
+ * @license Copyright (c) 2003-2018, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+/**
+ * @module engine/utils/bindtwostepcarettoattribute
+ */
+
+import { keyCodes } from '@ckeditor/ckeditor5-utils/src/keyboard';
+
+/**
+ * This helper adds two-steps caret movement behaviour for given attribute.
+ *
+ * When caret is moving by arrow keys and reach bound of text with attribute for which behaviour is enabled
+ * then typing does not expand this attribute. Additional arrow key press is needed to "enter" to the text
+ * and start typing with this attribute. The same is is for leaving this text.
+ *
+ * When behaviour is enabled for bold attribute and caret is just before the attribute element then pressing right arrow
+ * will move caret to the attribute element instead of moving after next character:
+ *
+ * 		<p>foo[]<strong>bar</strong>biz<p> `->` <p>foo<strong>[]foo</strong>barr<p>
+ *
+ * The same is for "leaving" text:
+ *
+ * 		<p>foo<strong>bar[]</strong>biz<p> `->` <p>foo<strong>bar</strong>[]biz<p>
+ *
+ * And when moving left:
+ *
+ * 		<p>foo<strong>bar</strong>[]biz<p> `<-` <p>foo<strong>bar[]</strong>biz<p>
+ * 		<p>foo<strong>[]bar</strong>biz<p> `<-` <p>foo[]<strong>bar</strong>biz<p>
+ *
+ * @param {module:core/editor/editor~Editor} editor The Editor instance.
+ * @param {module:utils/dom/emittermixin~Emitter} emitter The emitter to which this behavior should be added.
+ * @param {String} attribute Attribute for which behaviour will be added.
+ */
+export default function bindTwoStepCaretToAttribute( editor, emitter, attribute ) {
+	const model = editor.model;
```

Comment:
AFAICS, the editor isn't needed here – we only need to pass the view and the model to this function. Let's make this function engine-only.